src/backend/distributed/connection/locally_reserved_shared_connections.c

```diff
@@ -342,62 +364,114 @@ EnsureConnectionPossibilityForNodeList(List *nodeList)
+		 * means that the execution can use the same connection, so reservation
+		 * is not necessary.
 		 */
-		bool found = false;
-		ReservedConnectionHashEntry *hashEntry =
-			AllocateOrGetReservedConectionEntry(workerNode->workerName,
-												workerNode->workerPort,
-												userId, MyDatabaseId, &found);
+		return true;
+	}
 
-		if (found)
-		{
-			/*
-			 * We have already reserved a connection for this user and database
-			 * on the worker. We only allow a single reservation per
-			 * transaction block. The reason is that the earlier command (either in
-			 * a transaction block or a function call triggered by a single command)
-			 * was able to reserve or establish a connection. That connection is
-			 * guranteed to be avaliable for us.
-			 */
-			continue;
-		}
+	/*
+	 * We are trying to be defensive here by ensuring that the required hash
+	 * table entry can be allocated. The main goal is that we don't want to be
+	 * in a situation where shared connection counter is incremented but not
+	 * the local reserved counter due to out-of-memory.
+	 *
+	 * Note that shared connection stats operate on the shared memory, and we
+	 * pre-allocate all the necessary memory. In other words, it would never
+	 * throw out of memory error.
+	 */
+	bool found = false;
+	ReservedConnectionHashEntry *hashEntry =
+		AllocateOrGetReservedConectionEntry(workerNode->workerName,
+											workerNode->workerPort,
+											userId, MyDatabaseId, &found);
 
+	if (found)
+	{
+		/*
+		 * We have already reserved a connection for this user and database
+		 * on the worker. We only allow a single reservation per
+		 * transaction block. The reason is that the earlier command (either in
+		 * a transaction block or a function call triggered by a single command)
+		 * was able to reserve or establish a connection. That connection is
+		 * guranteed to be available for us.
+		 */
+		return true;
+	}
+
+	if (waitForConnection)
+	{
 		/*
 		 * Increment the shared counter, we may need to wait if there are
 		 * no space left.
 		 */
 		WaitLoopForSharedConnection(workerNode->workerName, workerNode->workerPort);
+	}
+	else
+	{
+		bool incremented =
+			TryToIncrementSharedConnectionCounter(workerNode->workerName,
+												  workerNode->workerPort);
+		if (!incremented)
+		{
+			/*
+			 * We could not reserve a connection. First, remove the entry from the
+			 * hash. The reason is that we allow single reservation per transaction
+			 * block and leaving the entry in the hash would be qualified as there is a
+			 * reserved connection to the node.
+			 */
+			bool foundForRemove = false;
+			hash_search(SessionLocalReservedConnections, hashEntry, HASH_REMOVE,
+						&foundForRemove);
+			Assert(foundForRemove);
 
-		/* locally mark that we have one connection reserved */
-		hashEntry->usedReservation = false;
+			return false;
+		}
 	}
+
+	/* locally mark that we have one connection reserved */
+	hashEntry->usedReservation = false;
```

Comment:
is there any potential for OOMs causing crashes in the logic surrounding SessionLocalReservedConnections?

Reply:
I think our approach is documented here: https://github.com/citusdata/citus/blob/2afb806e7ebdb9dd52b587537d59ea0b02b2877c/src/backend/distributed/connection/locally_reserved_shared_connections.c#L363-L377

Basically, we allocate entries on `SessionLocalReservedConnections` before we increment shared counter. If the allocation on `SessionLocalReservedConnections` fails, we would not have even touched the shared counters. If it succeeds, we should be good because the shared counters in on the shared memory, and we know that there are enough space already allocated.

Any further failures/OOMs after this point is handled in via [`DeallocateReservedConnections `](https://github.com/citusdata/citus/blob/2afb806e7ebdb9dd52b587537d59ea0b02b2877c/src/backend/distributed/connection/locally_reserved_shared_connections.c#L224-L261)  in `XACT_EVENT_COMMIT ` and `XACT_EVENT_ABORT `. So, they seem fine as well.

Given that this is a tricky part of the code, @marcocitus let me know if you have any further notes on this.